docs/decisions/0020-upstream-downstream.rst

```diff
@@ -0,0 +1,373 @@
+    on Open edX instances that are missing their linked upstream content.
+  * (Preserving a positive feature of the V1 LibraryContentBlock) The link
+    persists through export-import and copy-paste, regardless of whether the
+    upstream content actually exists. A "broken" link to upstream content is
+    seamlessly "repaired" if the upstream content becomes available again.
+
+* **Customizable:** On an OLX level, authors can still override the value
+  of any field for a piece of downstream content. However, we will empower
+  Studio to be more prescriptive about what authors *can* override versus what
+  they *should* override:
+
+  * We define a set of *customizable* fields, with platform-level defaults
+    like display_name and a max_attempts, plus the ability for external
+    XBlocks to opt their own fields into customizability.
+  * Studio may use this list to provide an interface for customizing
+    downstream blocks, separate from the usual "Edit" interface that would
+    permit them to make unsafe overrides.
+  * Furthermore, downstream content will record which fields the user has
+    customized...
+
+    * even if the customization is to simply clear the value of the fields...
+    * and even if the customization is made redundant in a future version of
+      the upstream content. For example, if max_attempts is customized from 3
+      to 5 in the downstream content, but the next version of the upstream
+      content also changes max_attempts to 5, the downstream would still
+      consider max_attempts to be customized. If the following version of the
+      upstream content again changed max_attempts to 6, the downstream would
+      retain max_attempts to be 5.
+
+  * Finally, the downstream content will locally save the upstream value of
+    customizable fields, allowing the author to *revert* back to them
+    regardless of whether the upstream content is actually available.
+
+* **Synchronizable, without surprises:** Downstream content can be *synced*
+  with updates that have been made to its linked upstream. This means that the
+  latest available upstream content field values will entirely replace all of
+  the downstream field values, *except* those which were customized, as
+  described in the previous item.
+
+* **Concrete, but flexible:** The internal implementation of upstream-downstream
+  syncing will assume that:
+
+  * upstream content belongs to a V2 content library,
+  * downstream content belongs to a course on the same instance, and
+  * the link is the stringified usage key of the upstream library content.
+
+  This will allow us to keep the implementation straightforward. However, we
+  will *not* expose these assumptions in the Python APIs, the HTTP APIs, or in
+  the persisted fields, allowing us in the future to generalize to other
+  upstreams (such as externally-hosted libraries) and other downstreams (such
+  as a standalone enrollable sequence without a course).
```

Comment:
What will happen if the upstream is not from a library but rather from the same course or from another course? (Say someone edits the OLX to put that.) Will we support that or throw an error?

Reply:
Yes, I agree :)

Reply:
Done